### ISSUE TITLE
Job not fired again after connection to database was lost

### ISSUE THREAD

**Issue by buschluschni** [lukas-krecan/ShedLock#1254]
Version: 4.42.0
Provider: JdbcTemplateLockProvider
Config: @EnableSchedulerLock(defaultLockAtMostFor = "10m", defaultLockAtLeastFor = "10m")

When connection to database is lost during a job run, it will not be triggered again, although the lock_until is in the past. 
After a restart (Spring Boot) it is triggered again as expected.

**Comment by lukas-krecan** [comment]
Hi, thanks for feedback. Is there anything you see in the log's? Stack traces?

**Comment by buschluschni** [comment]
No, I turned on DEBUG loglevel, there are no errors/stacktraces, just the debug message
_n.j.s.core.DefaultLockingTaskExecutor Not executing 'xxxx'. It's locked.
is written each time the job should be triggered.

**Comment by buschluschni** [comment]
It only happens when connection to database is temporarily unavailable during a job run. No matter how often the job was already executed.

**Comment by lukas-krecan** [comment]
From the code it can't happen 🙂 I can release a new version with more logging, would you be willing to help me to debug it?

**Comment by buschluschni** [comment]
I'd like to help you debug it 👍

**Comment by lukas-krecan** [comment]
Hi, I have just release version 4.43.0 with more logging, can you please try? Thank you

**Comment by buschluschni** [comment]
Hi, thanks for the new version. I tried it out, but I'm afraid the output won't help you much (Shedlock lock time is 1m, Scheduler runs every min):

```
2022-12-05 06:53:00.311 DEBUG 2628 --- [   scheduling-1] n.j.s.core.DefaultLockingTaskExecutor : Locked 'PriceInitJob', lock will be held at most until 2022-12-05T05:54:00.014Z
...
2022-12-05 06:55:00.004 DEBUG 2628 --- [   scheduling-1] n.j.s.core.DefaultLockingTaskExecutor : Not executing 'PriceInitJob'. It's locked.
...
2022-12-05 06:56:00.020 DEBUG 2628 --- [   scheduling-1] n.j.s.core.DefaultLockingTaskExecutor : Not executing 'PriceInitJob'. It's locked.
```

**Comment by lukas-krecan** [comment]
Can you please check what's in the ShedLock DB table (lock_until)?

**Comment by lukas-krecan** [comment]
Ok, the one hour difference is an interesting clue (unless it's a coincidence). Do you have `usingDbTime` in you lock provider config or not?

**Comment by lukas-krecan** [comment]
Can you please share some logs?

**Comment by devapeapod** [comment]
@lukas-krecan We didn't see any scheduler logs during that period. So far, it looks like a long-running query was blocking the shed_lock table update, and the other pod was just hanging around waiting to update shed_lock.

**Comment by lukas-krecan** [comment]
Hi, without the logs it's impossible to say what's going on. There are the following options

1. The lock was held for lockAtLeastFor and the fact that everything started to work after the pod restart is just a coincidence (the lock just timed-out). You can check the ShedLock table and see what lock are active after the DB restart.
2. Or the pod restart indeed helped. How this could happen? In my mind there are two main options. Either the JDBC connection pool got somehow corrupted or the Spring scheduler stopped working. In the former case you would see a lot of errors in the logs. I can imagine the latter case happening if, for example, you don't have timeouts configured correctly and all the threads in Spring scheduler got stuck. But as I said, it's impossible to tell without more info.

**Comment by ignusin** [comment]
@lukas-krecan hello, Sir.

I was going through similar issue with "Not executing. It's locked" message.
For some reason, sometimes (i still don't know why, and still investigating on my side), having a blank Oracle database - no lock record is getting created in the database.

I was investigating the source code of StorageBasedLockProvider, the following line https://github.com/lukas-krecan/ShedLock/blob/master/shedlock-core/src/main/java/net/javacrumbs/shedlock/support/StorageBasedLockProvider.java#L80 seems to be adding entry to in-memory cache even if no record was created in database.

That leads to situation, when no further attempts to insert record to the database are made.

I enabled trace logs for Shedlock, however no log entries for "Duplicate key" or "Unexpected exception" are present in the log file.

Could you, please, provide your thoughts on the issue observed?
Thank you.

**Comment by lukas-krecan** [comment]
Hi, thanks a lot, I think I have the cause. If ShedLock started and the DB was empty but not available, the record was erroneously inserted into lockRecordRegistry and the insert has not been retried.

**Comment by lukas-krecan** [comment]
@ignusin Just to double check, what LockProvider do you use? Plain JDBC or JdbcTemplate?

**Comment by lukas-krecan** [comment]
Just for future reference, this is what may be happening:

1. It should not be possible with JdbcTemplateLockProvider as it fails when creating TX for insert
2. It may be happening for JdbcLockProvider as it does not create TX and insert returns false even if the DB is not available
3. The assumption is that update fails on exception and [this catch block](https://github.com/lukas-krecan/ShedLock/blob/master/shedlock-core/src/main/java/net/javacrumbs/shedlock/support/StorageBasedLockProvider.java#L87) should fix it.
4. It's unlikely, but the DB can come online between insert and update and in that case the code will end-up in an incorrect state. (lockRecordRegistry will contain the record so the insert will never be attempted)

**Comment by ignusin** [comment]
@lukas-krecan we use JdbcTemplate, Sir.

**Comment by lukas-krecan** [comment]
I have refactored exception handling in 7.0.0 so it should work better now
